app/code/core/Mage/Payment/Model/Observer.php

```diff
@@ -32,7 +32,14 @@ public function salesOrderBeforeSave($observer)
         /** @var Mage_Sales_Model_Order $order */
         $order = $observer->getEvent()->getOrder();
 
-        if ($order->getPayment() && $order->getPayment()->getMethodInstance()->getCode() != 'free') {
+        if (($payment = $order->getPayment())
+            && ((
+                    Mage::helper('payment')->getMethodModelClassName($payment->getMethod()) !== null
+                    && $payment->getMethodInstance()->getCode() !== 'free'
+                )
+                || ($payment->getMethod() !== 'free')
+            )
+        ) {
```

Comment:
do we need such a complex if? to me it's extremely difficult to read, but also, in app/code/core/Mage/Adminhtml/Block/Sales/Order/View.php the "if" is so much simpler, shouldn't it work also here? or could we at least make it more readable?

Reply:
Thanks for your feedback! I think I've now found a much better solution that is also easier to read.
This new variant also has the advantage that now even the entire line for an order in `sales_flat_order_payment` can be removed from the database and the pages will still render.
In projects I work with anonymized dev databases (created with [gdpr-dump](https://github.com/Smile-SA/gdpr-dump)) with very few orders, in which the `sales_flat_order_payment` is truncated. With this new variant this is no longer a problem. :)

Could you please give it a try?